doc/adr/0002-steppable-cek-machine.md

```diff
@@ -0,0 +1,235 @@
+
+We currently decide to not take this approach because:
+
+- It's not clear whether we can implement it this way. Implementing a separate machine is simpler.
+- Some slowdown may still happen due to GHC optimizations being very unreliable.
+- It's difficult to detect slow downs due to this as we add more builtin functions.
+
+If the separate machine turns out to not provide correct debugging information, we may revisit this option again.
+
+## Decision: implementing it as a coroutine system
+
+This section describes the proposed implementation of the debugging machine.
+
+We first abstracts out the computation to "steps" on our current machine. We then implement a coroutine system to add the debugging functionalities.
+
+### Abstracting our the computation to "steps"
+
+This abstraction has been implemented in [PR#4909](https://github.com/input-output-hk/plutus/pull/4909/).
+
+The current machine inlined the steps. We separate each steps into separate functions. They all return a `CekState`:
+
+```haskell
+data CekState uni fun =
+    -- the next state is computing
+    Computing WordArray (Context uni fun) (Closure uni fun)
+    -- the next state is returning
+    | Returning WordArray (Context uni fun) (CekValue uni fun)
+    -- evaluation finished
+    | Terminating (Term NamedDeBruijn uni fun ())
+
+data Closure uni fun = 
+  Closure (Term NamedDeBruijn uni fun ()) (CekValEnv uni fun)
+```
+
+The computing step is `computeCekStep` with the following signature:
+
+```haskell
+computeCekStep
+    :: forall uni fun s
+    . (Ix fun, PrettyUni uni fun, GivenCekReqs uni fun s)
+    => WordArray
+    -> Context uni fun
+    -> Closure uni fun
+    -> CekM uni fun s (CekState uni fun)
+```
+
+Similarly for the returning step (`returnCekStep`). Then we link up all the steps with `continue`, and the machine behaves very similar to our current one:
+
+```haskell
+continue :: forall uni fun s
+    . (Ix fun, PrettyUni uni fun, GivenCekReqs uni fun s)
+    => CekState uni fun
+    -> CekM uni fun s (Term NamedDeBruijn uni fun ())
+continue (Computing !unbudgetedSteps ctx (Closure term env)) = do
+    state <- computeCekStep unbudgetedSteps ctx env term
+    continue state
+continue (Returning !unbudgetedSteps ctx val) = do
+    state <- returnCekStep unbudgetedSteps ctx val
+    continue state
+continue (Terminating term) = pure term
+```
+
+### Alternative: use CPS
+
+This approach has been suggested to improve the performance of the debugging machine. The abstraction proposed above does not have good performance. To improve performance, we need to get rid of the creation of the state objects. This is because GHC probably doesn't have the capability to eliminate those, because the place where they're destructed is too far from where they're constructed. That means that instead we probably need to do something with continuations:
```

Comment:
> This approach has been suggested to improve the performance of the debugging machine. The abstraction proposed above does not have good performance.

For the production machine. For the debugging machine the performance of this approach is perfectly OK and we don't care about performance here much anyway.

Reply:
Exactly. This is only relevant if we're talking about refactoring the production machine to support both ways, then we need to make it fast. So it's another alternative alongside "polymorphic compute/return steps"